### PR TITLE
fix OADP-5044 and other VSL validation errors

### PR DESCRIPTION
**Pull request by stillalearner**
## Why the changes were made

Fixed some VSL validation errors like:

-> throwing error on incorrect provider in VSL
-> throwing error on missing region in AWS VSL
-> throwing error if missing plugins for AWS, GCP, Azure VSLs

## How to test the changes made

Use "make deploy-olm" and configure VSL in DPA spec.

1. Provide incorrect provider in VSL
2. Configure AWS VSL, and dont provide region in VSL spec.
3. Add AWS/GCP/Azure VSL and dont add relevant velero plugin.

### CODE REVIEW

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:125]: Here, lets delete `vsl` variable (it was used only by `EventRecorder`) and use same approach as the other function in this file does to get VSL name

https://github.com/openshift/oadp-operator/blob/fb2411e2948874a6abe61ca31b1cd9957621ebb6/controllers/vsl.go#L227-L231

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:133]: I think we should remove these logs, OR, change them to be level error, since we are erroring it out now

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:135]: I really think we should start to give more verbose error messages to users, as I explain here https://github.com/openshift/oadp-operator/issues/1536

My suggestion for here, for example, would be
> DPA spec.snapshotLocations[**index**].velero.provider is invalid: **wrong-one** is not a valid value. Use of of **options**

File: controllers/validator_test.go
Comment (mateusoliveira43) [inline controllers/validator_test.go:1041]: Looking this code, I think this is not a valid change
https://github.com/openshift/oadp-operator/blob/fb2411e2948874a6abe61ca31b1cd9957621ebb6/controllers/bsl.go#L50-L68
validation should accept both `aws` and `velero.io/aws` (but this should be fixed in the future, we can open an issue to discuss only allowing `aws`), right maintainers?

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:149]: here message is wrong, no?

It is referencing the `key` not the `value`, in formatting

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:156]: we do not have this kind of validation for BSL, right? Should we add it (if yes, another PR)

**Review by mateusoliveira43 — CHANGES_REQUESTED** [review]
Added some comments, but not much changes needed for approval :smile:

File: controllers/validator_test.go
Comment (stillalearner) [reply controllers/validator_test.go:1041]: @mateusoliveira43 I have fixed this intentionally, for now , for this test to pass.

Because this is a bug here actually, there are multiple issues regarding this, which i plan to open soon.

Then, we can separately fix this issue.

File: controllers/vsl.go
Comment (stillalearner) [reply controllers/vsl.go:125]: ack

File: controllers/vsl.go
Comment (stillalearner) [reply controllers/vsl.go:149]: ack

File: controllers/vsl.go
Comment (stillalearner) [reply controllers/vsl.go:135]: Is the name of VSL needed then?

File: controllers/vsl.go
Comment (stillalearner) [reply controllers/vsl.go:156]: hmm, probably not, will confirm and create a PR if not. Noted.

File: controllers/vsl.go
Comment (mateusoliveira43) [reply controllers/vsl.go:135]: I think just yaml path is fine (and most of the time users do not even put name on those objects, right? so it will not be a valuable information for them) 

With yaml path I would say error message is good enough, because if user does not understand it, we will understand it immediately and can easily help the user fix the problem

Comment (stillalearner) [reply controllers/vsl.go:135]: ya sounds good. Thanks

File: controllers/vsl.go
Comment (stillalearner) [reply controllers/vsl.go:133]: ack

File: controllers/validator_test.go
Comment (mateusoliveira43) [reply controllers/validator_test.go:1041]: Since this changes behavior (prior to this PR, users could specify `aws` or `velero.io/aws` in VSL, and after it, users will be able to only specify `aws`) I think is important to discuss this prior to merge

(easy solution could be that if user specifies `velero.io/aws`, the error message tell it to use `aws` instead, but we need to verity our deprecation policy as well, maybe first we need to deprecate it)

Comment (stillalearner) [reply controllers/validator_test.go:1041]: no, no, this PR does not change any behaviour, i have only modified the test case to use "aws" for sometime.
Earlier, velero.io/aws was also creating a Warning event, and was getting ignored.

The reason it is failing now with "velero.io" prefix is a different bug, which is causing issue now. I will raise a thread in slack. That needs to be fixed first, and after that, we can have any value here.

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:122]: I would use variables inside error messages as much as possible

For example, creating these variables inside for loop (after line 112)
```go
vslYAMLPath := fmt.Sprintf("spec.snapshotLocations[%v]", i)
veleroVSLYAMLPath := vslYAMLPath + ".velero"
```
So if these paths update in the future, it is easy to update error messages (only one place, instead of every message individually)

And then changing this message to
```go
return false, fmt.Errorf("DPA %v.provider %s is invalid: only %v are supported", veleroVSLYAMLPath, vslSpec.Velero.Provider, getVSLPluglins())
```
(`getVSLPluglins` does not exist, created for the example)

What you think?

Comment (stillalearner) [reply controllers/vsl.go]: ya, variable part is good.

Comment (stillalearner) [reply controllers/vsl.go]: changes pushed. Thanks you

File: controllers/vsl.go
Comment (mateusoliveira43) [reply controllers/vsl.go:123]: only this log was left inside `ValidateVolumeSnapshotLocations` function, do not forget to remove it :)

Comment (stillalearner) [reply controllers/vsl.go]: oops. done.

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:123]: lets use `AWSProvider`, `GCPProvider` and `AzureProvider` variables here in the message

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:130]: Typo in `config` (better to be lowercase, matching json tag, so commands like `oc get dpa velero-sample -n openshift-adp -o=jsonpath="{.spec.snapshotLocations[0].velero.config}"` work)

File: controllers/vsl.go
Comment (mateusoliveira43) [inline controllers/vsl.go:143]: better to use imperative form here, no?

would also add path to aws plugin config in DPA

Suggestion 

> to use VSL for AWS specified in DPA spec.snapshotLocations[index], AWS plugin must be present in spec.configuration.velero.defaultPlugins

**Review by shubham-pampattiwar — CHANGES_REQUESTED** [review]
@stillalearner Please add specific unit tests for the validation error case scenarios related to this PR:
-> throwing error on incorrect provider in VSL
-> throwing error on missing region in AWS VSL
-> throwing error if missing plugins for AWS, GCP, Azure VSLs

**Comment by stillalearner** [comment]
> @stillalearner Please add specific unit tests for the validation error case scenarios related to this PR: -> throwing error on incorrect provider in VSL -> throwing error on missing region in AWS VSL -> throwing error if missing plugins for AWS, GCP, Azure VSLs

Sure, Thanks! All scenarios covered now.

File: controllers/vsl_test.go
Comment (mateusoliveira43) [inline controllers/vsl_test.go:1]: For follow up (you already did an amazing job in this PR @stillalearner!!!)

As I explain in https://github.com/openshift/oadp-operator/issues/1536 these tests do not gives us confidence, because we do not check error messages (only if function errors out). So some tests here may be currently wrong

**Review by mateusoliveira43 — APPROVED** [review]

File: controllers/vsl_test.go
Comment (stillalearner) [reply controllers/vsl_test.go:1]: Thanks @mateusoliveira43 !

Yes, for the error messages, dont worry, Its getting ready in another PR as part of 1536 fixes.

**Review by kaovilai — APPROVED** [review]
/lgtm

**Pull request merged**
